geograypher/utils/image.py

```diff
@@ -19,3 +24,142 @@ def get_GPS_exif(filename):
     )
     # https://stackoverflow.com/questions/54196347/overwrite-gps-coordinates-in-image-exif-using-python-3-6
     return lon, lat
+
+
+def perspective_from_equirectangular(
+    equi_img: np.ndarray,
+    fov_deg: float,
+    output_size: Tuple[float] = (1440, 1440),
+    yaw_deg: float = 0,
+    pitch_deg: float = 0,
+    roll_deg: float = 0,
+    warp_order: int = 1,
+    oversample_factor: int = 1,
+    return_mask: bool = False,
+) -> Tuple[np.ndarray, Optional[np.ndarray]]:
+    """
+    Sample a perspective image from an equirectangular (360) image. Different parameters of the
+    "virtual camera" can be controlled, including the field of view, oreintation (yaw, pitch, roll),
+    and output image size.
+
+    Args:
+        equi_img (np.ndarray):
+            Equirectangular image in (H, W, C) format. Currently only float (0,1) images are
+            supported but this is a TODO.
+        fov_deg float: camera horizontal field of view.
+        output_size Tuple[float]: Shape of the output image to sample (i, j).
+        yaw_deg (float, optional): yaw camera orientation. Defaults to 0.
+        pitch_deg (float, optional): pitch camera orientation. Defaults to 0.
+        roll_deg (float, optional): roll camera orientation. Defaults to 0.
+        warp_order (int, optional): Interpolation order to use when sampling pixels. Defaults to 1.
+        oversample_factor (int, optional):
+            Sample this number times more pixels in each dimension. This helps avoid aliasing but
+            increases runtime. Defaults to 1.
+        return_mask (bool, optional):
+           Return a mask showing what pixels from the original image were
+             sampled. Defaults to False.
+
+    Returns:
+        np.ndarray: The sampled perspective image
+        Optional[np.ndarray]]: The mask of sampled locations, if requested
+    """
+    # Convert an equirectangular (360) image to a perspective view.
+    H, W = equi_img.shape[:2]
+
+    # Output image dimensions
+    out_w, out_h = output_size
+
+    # Sample a larger image and then downsample to reduce aliasing
+    out_w = int(out_w * oversample_factor)
+    out_h = int(out_h * oversample_factor)
+
+    # FOV and angles in radians
+    fov = np.deg2rad(fov_deg)
+    yaw = np.deg2rad(yaw_deg)
+    pitch = np.deg2rad(pitch_deg)
+    roll = np.deg2rad(roll_deg)
+
+    # Compute the aspect ratio of the requested image dimensions
+    aspect_ratio = out_h / out_w
+
+    # Create homogenous image coordinates for the output image
+    x = np.linspace(-np.tan(fov / 2), np.tan(fov / 2), out_w)
```

Comment:
Nit - also just a style thing (feel free to ignore), but if you wanted to save np.tan(fov / 2) as a short variable it could shorten these lines and avoid copies